### ISSUE TITLE
IUS BBY addy intro as nice test case for ADLIB sound and RTC interrupts

### ISSUE THREAD

**Issue by mcericdos** [stsp/dosemu2#395]
The addy / intro IUS might be an interesting test case:

It is small, has ADLIB sound and RTC IRQ timed sound and 3d starfield animation with motion lag effect... So you can use it to test the quality of your sound and RTC emulation :-)

[ius-bbs-addy-intro.zip](https://github.com/stsp/dosemu2/files/1082517/ius-bbs-addy-intro.zip)

...

I have two versions of the binary, but do not remember why. The difference is one byte:

Original: 587: 78F1  JS  057A

Patched: 587: 7800 JS $+2 (in other words, make the conditional jump a no-op)

The patch has the effect that IUS no longer modifies activities when it finds the MSB of GDT[5] to be set. Which means that the GDT NUL segment (which is never used but still expected by some emulators?) is a "present" segment.

...

Maybe the unpatched version skipped some bells and whistles when run inside the DOS box of Windows or when EMM386 is active or something similar to that. Feel free to use either of the two versions for checking your ADLIB and RTC IRQ simulation quality :-)

Code areas in terms of IP offsets in the COM: 100 to 104, 128 to 68F

Data areas: 105 to 127 string 04, "*** THX TO FUTURE CREW! ***", e8, 03, 00

... and the actual data are at 690 to 903 (followed by a bunch of boring 00 bytes)

**Comment by stsp** [comment]
Please name exactly the problems and
artefacts that you see with it, and put them
into a ticket's description too please.

**Comment by mcericdos** [comment]
In Dosemu2 2.0pre4, the unpatched version only showed a text mode message and exits. 

The patched version takes a lot of CPU but only a few pixels of the star field are actually shown (like 10 pixels near the upper left corner) and are not properly animated and there is no sound.

I would like to **ask a reader who has the newest version of Dosemu2 installed** whether IUS fully works there. 

If yes, then IUS is useful as regression test in the future.

If no, then IUS can be used as test case for the following feature suggestions:

- Make the NUL segment in the GDT an actual non-present NUL segment
- Improve the support for Adlib sound
- Improve the support for interrupts generated by the RTC, here: periodic IRQ
- Possibly improve the VGA emulation if that has a problem after fixing the other aspects

Of course it is okay if RTC interrupts get pooled into bursts when the Linux timer tick is slower than the simulated RTC periodic IRQ. I predict jitter to be sufficiently small both in the animation and in the sound.

I hope somebody with the newest Dosemu2 can give it a quick try. The BBS addy started within seconds on real hardware with real ISA soundblaster and shows a scrolling text made from large pixels flying through a 3d star field with some simple background music. The camera is rotating around it and a simple motion blur style effect shows a few pictures in different white / blue shades, possibly using a few bits per pixel for each time step. In other words, this intro presents an interesting challenge to emulators and is entertaining to run :-)

Unfortunately I no longer have sufficiently compatible sound hardware, in spite of having a small collection of PCI sound cards with some degree of hardware soundblaster imitation support. At some point, PCI slots and chipsets apparently became worse in cooperating with such tweaks.

Sorry about not testing this myself in the newest Dosemu2 first: Compiling it with sensible settings is something that I do not have experience with yet, so I would prefer to get a judgement about current IUS compatibility from somebody who has a well-compiled and fresh Dosemu2 sitting on their disk already anyway.

**Comment by stsp** [comment]
OK, I have no idea what 2 versions
you have, but I take the only version
is there in the archive you pointed,
and it works perfectly with video and
music, thanks.

**Comment by stsp** [comment]
If you can't compile dosemu2, please use
the daily builds from here:
https://code.launchpad.net/~dosemu2/+recipe/dosemu2-daily

**Comment by mcericdos** [comment]
Cool that there are daily builds, but I have problems installing them: They would overwrite /usr/share/fonts/X11/misc/vga11x19.pcf.gz already is part of bitchx...

When I first remove bitchx, then I can install dosemu2. I cannot re-install bitchx afterwards, though, as it got dropped from newer repositories of my distro. Bad luck, I liked bitchx, but I like dosemu more ;-)

Interestingly, dosemu2 2.0pre6.1 initially did not work with IUS, but fixing a misconfigured COM1 (I had still configured it to use my no longer existing ttyS1 to keep DOS and Linux serial port tinkering separated) made IUS2 fully work in "dosemu -X" :-) In spite of two warnings: libSDL2-2.0.so.0 is not found and fluidsynth sample data RAM cannot be pinned.

So the only remaining issue is that the GDT NUL segment is not a true NUL segment but rather a segment which has the "present" bit set, which makes the unpatched version of IUS bail out and show only the text without graphical animation and sound.

I could imagine that changing that could have bad side effects on yet other software for DOS, but would be happy to hear your expert advice about that.

**Comment by jschwartzenberg** [comment]
Just in case here is the attachment running on directly on hardware:
https://goo.gl/photos/iKM5mS38USGGxyPf8

**Comment by mcericdos** [comment]
Is it just my browser or is the sound (as sort of expected) absent on your (probably too modern, like mine) hardware?

**Comment by stsp** [comment]
> Cool that there are daily builds, but I have problems installing them:
> They would overwrite /usr/share/fonts/X11/misc/vga11x19.pcf.gz already is part of bitchx...

This is an ubuntu deficiency actually, as it
still does not support /etc/X11/fontpath.d:
https://github.com/stsp/dosemu2/issues/364#issuecomment-296429949
The simplest work-around is to probably
revert this commit of @andrewbird:
https://github.com/stsp/dosemu2/pull/371/commits/b1ed8de5c101305ee9f19d1fc150fc5bb9140f9e
While this is definitely not the right solution,
I can't think of anything else unless ubuntu
to start supporting new features in a timely
fashion instead of trying to work around them
with a bunch of debhelper scripts.
Andrew,  can you suggest anything here?

> In spite of two warnings: libSDL2-2.0.so.0 is not found

Could you please do
`ldd libplugin_sdl.so` in a directory where `libplugin_sdl.so`
of dosemu2 package is installed? I think you took
the package for the different ubuntu version, and
so even though it got installed, some libs took
different names.

> and fluidsynth sample data RAM cannot be pinned.

Your fluidsynth is too old, but the new one is not
packaged because the upstream have died.
dosemu2 already has the rip-offs from the
fluidsynth git with a lot of manual patching,
but I tried to keep them small rather than to
fork the entire code-base. For now you can
build fluidsynth from git and then dosemu2
will glue it with its own rip-offs to get something
functional. Or you can just ignore that message
on startup. :)

> I could imagine that changing that could have bad
> side effects on yet other software for DOS, but would
> be happy to hear your expert advice about that.

Long story.
Currently it is not possible to change because the
default cpuemu backend we use is kvm, where
you can't change the behaviour of SGDT easily.
The upcoming intel CPUs (icelake) are going to
support the feature called UMIP that allows you
to do exactly this.
But if you switch to the emulation backend, then
you can see the `ERROR: SGDT not implemented`.
This one can be easily converted into anything you
want, if you do that yourself, of course. :) Just
create a buffer with fake GDT and make the cpuemu
to return the pointer to it with sgdt.

**Comment by mcericdos** [comment]
Hmm pity to hear about the sort-of-demise of fluidsynth...

Regarding SDL support: I used the DEB for exactly my version of Ubuntu, but I simply did not have SDL2 installed. I have the impression that Dosemu2 deliberately does not REQUIRE it. After installing SDL2 (libsdl2-dev & libsdl2-dgb, with apt-get, with all automatic dependencies of that), Dosemu2 no longer shows the warning. Sound and graphics work equally well before and after installing SDL2.

So I would say it is a good idea that Dosemu2 makes installation of SDL2 optional, although it could probably contain a recommendation flag for SDL2. Unless of course there are features for which SDL2 works better than normal dosemu -X mode...?

**Comment by jschwartzenberg** [comment]
>Is it just my browser or is the sound (as sort of expected) absent on your (probably too modern, like mine) hardware?

Be sure to click on the image as initially only an animation is shown, not the actual video. You should hear sound.

**Comment by mcericdos** [comment]
Indeed, now there is sound. Brave new google world. Which type of sound hardware do you use?

**Comment by jschwartzenberg** [comment]
It's an SB AWE64.

**Comment by stsp** [comment]
> So I would say it is a good idea that Dosemu2 makes installation of SDL2 optional

Ah, yes, indeed... sometimes I forget what I
did myself recently - not a good sign. :) So
dosemu2 only build-requires it, but you haven't
built it, so you got this.
But do you really need `-X` or does it fall-back
to X automatically?

> Unless of course there are features for which SDL2 works better than normal dosemu -X mode...?

https://github.com/stsp/dosemu2/issues/394
But the opposite is also true, for example full-screen
support with SDL is mostly broken.

**Comment by mcericdos** [comment]
SB AWE64 exists in different versions: True ISA and a PCI version, the latter probably shipping with a driver similar to SB PCI and SB Live, where the driver creates an emulation of Soundblaster compatible hardware and works only for games which use either real mode or driver-compatible DOS extenders...

Stas, I do not need -X, but I wanted to say that explicitly using either SDL or X made no difference in how well IUS works in a window :-) Luckily I have not attempted to use SDL full screen mode yet ;-)

**Comment by jschwartzenberg** [comment]
I was not aware there was also a PCI version of the SB AWE64. It's an ISA card in my case. I'm indeed not aware of any good option to get proper SB/OPL3 hardware support in dos on a modern system.

**Comment by stsp** [comment]
Please note that SDL is not only used for
video, but also for sound. This is the reason
you get an error even when explicitly
specifying `-X`. This is now fixed.

**Comment by andrewbird** [comment]
> This is an ubuntu deficiency actually, as it still does not support /etc/X11/fontpath.d: #364 (comment)
The simplest work-around is to probably revert this commit of @andrewbird:

I looked at the existing packaging for Dosemu1 in Debian/Ubuntu and depending upon which version there seems to be a variation of how to install the fonts
- install fonts in a separate package called xfonts-dosemu, then have Dosemu use 'Recommends' to maybe pull it in.
- Install the fonts in '/usr/X11R6/lib/X11/fonts/misc/dosemu' maybe it will still work with debhelper, but I don't know and didn't try it. Seems to be the way Debian Jessie and Ubuntu Zesty do it now.
- Install the conflicting font with different name '/usr/share/fonts/X11/misc/dosemu-vga11x19.pcf.gz', apparently this was done in Debian itself in 2006, as mentioned [here](https://bugs.launchpad.net/ubuntu/+source/dosemu/+bug/54431)

Finally i tried 'apt-file update && apt-file search vga11x19.pcf.gzvga11x19.pcf.gz' and got no hits other than my locally installed dosemu2. As @mcericdos mentioned, BitchX seems to be no longer in the official repos and no other standard package provides that font, at least for Ubuntu 16.04. I also checked the current BitchX git repo on sourceforge and it does not contain the font. Its help file shows the font should be obtained from http://personal.nbnet.nb.ca/mudhoney/fonts/vga11x19.pcf but that seems to be a dead link.
So in summary I think it's best to leave as is, there's no conflict with an official package in Debian or Ubuntu, the old bitchx .deb is no more or less correct than us in claiming the right to install the file there.

**Comment by stsp** [comment]
OKey, thanks.
Lets do exactly that.
But if the next time this pops up,
lets revert to a separate dir. :)
